compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -2767,8 +2767,28 @@ abstract class GenJSCode extends plugins.PluginComponent
 
       val returnCount = countsOfReturnsToMatchEnd.remove(matchEndSym).get
 
-      genOptimizedLabeled(encodeLabelSym(matchEndSym), toIRType(matchEnd.tpe),
-          translatedCases, returnCount)
+      val LabelDef(_, List(matchEndParam), matchEndBody) = matchEnd
+
+      val innerResultType = toIRType(matchEndParam.tpe)
+      val optimized = genOptimizedLabeled(encodeLabelSym(matchEndSym),
+          innerResultType, translatedCases, returnCount)
+
+      matchEndBody match {
+        case Ident(_) if matchEndParam.symbol == matchEndBody.symbol =>
+          // matchEnd is identity.
+          optimized
+
+        case Literal(Constant(())) =>
+          // Unit return type.
+          optimized
```

Comment:
These two cases probably call for AST tests that check we're not emitting `VarDef`s when it is not necessary.